File: android/src/main/kotlin/com/ovaom/flutter_libpd/pd_controller/pd_patch_file_manager/PdPatchFileManagerImpl.kt
package com.ovaom.flutter_libpd.pd_controller.pd_patch_file_manager

import org.puredata.core.PdBase
import java.io.IOException

class PdPatchFileManagerImpl : PdPatchFileManager {
    private var openedFiles: MutableList<Pair<PatchFile, Int>> = mutableListOf()

    override val openedPatches: List<PatchFile>
        get() = openedFiles.toList().map { it.first }

    override fun openPatch(file: PatchFile) {
        if (isPatchOpened(file)) {
            return
        }
        try {
            val patchId = PdBase.openPatch(file.fullPath)
            openedFiles.add(Pair(file, patchId))
        } catch (error: IOException) {
            throw  PatchFileError.OpenFailed(file)
        }
    }

    override fun closePatch(file: PatchFile) {
        findPatchIndex(file)?.let {
            PdBase.closePatch(openedFiles[it].second)
            openedFiles.removeAt(it)
        }
    }

    private fun isPatchOpened(file: PatchFile): Boolean {
        return findPatchIndex(file) != null
    }

    private fun findPatchIndex(file: PatchFile): Int? {
        val index = openedFiles.indexOfFirst { it.first == file }
        return if (index != -1) index else null
    }
}

File: ios/Classes/PdController/PdPatchFileManager/PdPatchFileManagerImpl.swift
//
//  PdPatchFileManagerImpl.swift
//  flutter_libpd
//
//  Created by flav on 25/05/2022.
//

import Foundation
import libpd_ios

public class PdPatchFileManagerImpl: PdPatchFileManager {

    public init() {}
    
    private var openedFiles: [(PatchFile, PdFile)] = []
    
    private func findPatchIndex(_ file: PatchFile) -> Int? {
        openedFiles.firstIndex(where: { $0.0 == file })
    }
    
    private func isPatchOpened(_ file: PatchFile) -> Bool {
        findPatchIndex(file) != nil
    }
    
    public var openedPatches: [PatchFile] {
        openedFiles.map { $0.0 }
    }
    
    public func openPatch(_ file: PatchFile) throws {
        if isPatchOpened(file) { return }
        guard let pdFile = PdFile.openNamed(file.name, path: file.path) else {
            throw PatchFileError.openFailed(file)
        }
        openedFiles.append((file, pdFile))
    }
    
    public func closePatch(_ file: PatchFile) {
        guard let index = findPatchIndex(file) else { return }
        openedFiles[index].1.close()
        openedFiles.remove(at: index)
    }
}
